src/app/modules/react-migration/react-migration.controller.ts

```diff
@@ -0,0 +1,118 @@
+import path from 'path'
+
+import { FormAuthType } from '../../../../shared/types'
+import config from '../../config/config'
+import { ControllerHandler } from '../core/core.types'
+import * as FormService from '../form/form.service'
+import * as PublicFormController from '../form/public-form/public-form.controller'
+import { RedirectParams } from '../form/public-form/public-form.types'
+import * as HomeController from '../home/home.controller'
+
+const RESPONDENT_COOKIE_NAME = 'v2-respondent-ui'
+const ADMIN_COOKIE_NAME = 'v2-admin-ui'
+
+export type SetEnvironmentParams = {
+  ui: 'react' | 'angular'
+}
+
+export const RESPONDENT_COOKIE_OPTIONS = {
+  httpOnly: true,
+  sameSite: 'strict' as const,
+  secure: !config.isDev,
+}
+
+export const ADMIN_COOKIE_OPTIONS = {
+  httpOnly: true,
+  maxAge: 365 * 24 * 60 * 60, // 1 year
+  sameSite: 'strict' as const,
+  secure: !config.isDev,
+}
+
+const serveFormReact: ControllerHandler = (_req, res) => {
+  const reactFrontendPath = path.resolve('dist/frontend')
+  res.sendFile(path.join(reactFrontendPath, 'index.html'))
+}
+
+const serveFormAngular: ControllerHandler<
+  RedirectParams,
+  unknown,
+  unknown,
+  Record<string, string>
+> = (req, res, next) => {
+  PublicFormController.handleRedirect(req, res, next)
+}
+
+export const serveForm: ControllerHandler<
+  RedirectParams,
+  unknown,
+  unknown,
+  Record<string, string>
+> = async (req, res, next) => {
+  const formResult = await FormService.retrieveFormKeysById(req.params.formId, [
+    'authType',
+  ])
+  let showReact: boolean | undefined = undefined
+  let isSPCP = true
+
+  if (!formResult.isErr()) {
+    isSPCP = [FormAuthType.SP, FormAuthType.CP, FormAuthType.MyInfo].includes(
+      formResult.value.authType,
+    )
+  }
+
+  const threshold = isSPCP
+    ? config.reactMigrationConfig.respondentRolloutSPCP
+    : config.reactMigrationConfig.respondentRolloutNoSPCP
+
+  if (threshold <= 0) {
+    // Check the rollout value first, if it's 0, react is DISABLED
+    // And we ignore cookies entirely!
+    showReact = false
+  } else if (req.cookies) {
+    if (ADMIN_COOKIE_NAME in req.cookies) {
```

Comment:
maybe add comment here about priorityl let admins be dogfooder!

Reply:
https://github.com/opengovsg/FormSG/pull/3750/commits/3c5321d41c45bda994f562165d5e5f80634e9737